File: ios/UGames/ViewModels/HomeViewModel.swift
import Foundation
import Combine

@MainActor
final class HomeViewModel: ObservableObject {
    @Published private(set) var isLoading: Bool = false
    @Published private(set) var error: String?
    @Published private(set) var hero: Game?

    @Published private(set) var feedRecent: [Game] = []
    @Published private(set) var favoritesRow: [Game] = []
    @Published private(set) var spotlight: SpotlightBlock?
    @Published private(set) var genreRows: [GenreRow] = []
    @Published private(set) var profile: UserProfile?

    private let service: CatalogService
    private let favorites: FavoritesStore
    private var cancellables = Set<AnyCancellable>()
    private var loaded = false

    private let categoryRowLimit = 6

    init(service: CatalogService, favorites: FavoritesStore) {
        self.service = service
        self.favorites = favorites
        favorites.$games
            .receive(on: RunLoop.main)
            .sink { [weak self] g in self?.favoritesRow = Array(g.prefix(12)) }
            .store(in: &cancellables)
        service.$profile
            .receive(on: RunLoop.main)
            .sink { [weak self] p in
                guard let self = self else { return }
                let wasAnon = self.profile?.isAuthorized != true
                self.profile = p

                if wasAnon && p?.isAuthorized == true && self.loaded {
                    Task { await self.refreshFeedOnly() }
                }
            }
            .store(in: &cancellables)

        service.$gameSessionEndCount
            .dropFirst()
            .receive(on: RunLoop.main)
            .sink { [weak self] _ in
                guard let self = self, self.loaded else { return }
                Task { await self.refreshFeedOnly() }
            }
            .store(in: &cancellables)
    }

    private func refreshFeedOnly() async {
        do {
            let main = try await service.fetchFeedWithBlocks()
            digest(main: main)
            let categories = (try? await service.fetchCategories()) ?? []
            await fanOutGenreRows(categories: categories, exclude: main.flatGames.first?.appId)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func loadInitialIfNeeded() async {
        if loaded { return }
        loaded = true
        await refresh()
    }

    func refresh() async {
        isLoading = true
        error = nil
        defer { isLoading = false }
        do {
            let main = try await service.fetchFeedWithBlocks()
            digest(main: main)

            let categories = (try? await service.fetchCategories()) ?? []
            await fanOutGenreRows(categories: categories, exclude: main.flatGames.first?.appId)
        } catch {
            self.error = error.localizedDescription
        }

        let svc = service
        Task.detached(priority: .userInitiated) {
            await svc.refreshProfile()
        }
    }

    func toggleFavorite(_ game: Game) {
        favorites.toggle(game)
    }

    private func digest(main: FeedWithBlocks) {
        func isPromo(_ b: FeedBlock) -> Bool {
            b.type.compare("promo", options: .caseInsensitive) == .orderedSame
                || b.type.compare("adv", options: .caseInsensitive) == .orderedSame
        }
        let heroBlock = main.blocks.first(where: { !isPromo($0) && !$0.items.isEmpty })
        hero = heroBlock?.items.first ?? main.flatGames.max(by: { $0.ratingCount < $1.ratingCount })

        let spotlightBlock = main.blocks.first(where: {
            !isPromo($0) && $0 != heroBlock && $0.items.count >= 5
        })
        spotlight = spotlightBlock.map {
            SpotlightBlock(title: $0.title.isEmpty ? "Featured" : $0.title, games: $0.items)
        }

        feedRecent = Array(main.recentGames.prefix(12))

        if let hb = heroBlock, hb.items.count > 1 {
            let rest = Array(hb.items.dropFirst())
            genreRows = [GenreRow(title: "Fresh today", categoryName: nil, games: rest)]
        } else {
            genreRows = []
        }
    }

    private func fanOutGenreRows(categories: [GameCategory], exclude heroAppId: Int64?) async {
        let pick = Array(categories.prefix(categoryRowLimit))
        guard !pick.isEmpty else { return }

        let initial = genreRows
        var rows: [String: GenreRow] = [:]

        await withTaskGroup(of: (String, GenreRow?).self) { group in
            for cat in pick {
                group.addTask { [weak self] in
                    guard let self = self else { return (cat.name, nil) }
                    let resp = try? await self.service.fetchFeedWithBlocks(tab: cat.name)
                    let items = (resp?.flatGames ?? []).filter { $0.appId != heroAppId }
                    if items.isEmpty { return (cat.name, nil) }
                    return (cat.name, GenreRow(title: cat.title, categoryName: cat.name, games: Array(items.prefix(15))))
                }
            }
            for await (name, row) in group {
                if let row = row { rows[name] = row }
            }
        }

        let ordered = pick.compactMap { rows[$0.name] }
        genreRows = initial + ordered
    }
}

File: android/app/src/main/kotlin/games/yandex/wrap/ui/home/HomeViewModel.kt
package games.yandex.wrap.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import games.yandex.wrap.catalog.CatalogRepository
import games.yandex.wrap.catalog.FeedBlock
import games.yandex.wrap.catalog.FeedWithBlocks
import games.yandex.wrap.catalog.GameCategory
import games.yandex.wrap.catalog.Game
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Backs HomeScreen. Loads:
 * 1. main feed → hero + spotlight + a "fresh today" suggested row,
 * 2. server-side `recentGames` → Continue row (overrides local recents),
 * 3. top N categories from SSR + per-category feed → genre rows.
 *
 * Yandex's JSON feed only ever returns 1–4 untitled `suggested` blocks for
 * mobile platforms, so the only way to get distinct, titled rows on Home
 * is to fan out one feed call per top category. Done in parallel once and
 * cached for the session.
 */
class HomeViewModel(private val repository: CatalogRepository) : ViewModel() {

    private val _state = MutableStateFlow(HomeUiState())
    val state: StateFlow<HomeUiState> = _state.asStateFlow()

    val recent: StateFlow<List<Game>> = repository.recentGames()
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    val favorites: StateFlow<List<Game>> = repository.favoritesAsGames()
        .stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    private val categoryRowLimit = 6

    init {
        refresh()
        viewModelScope.launch {
            combine(recent, favorites) { r, f -> r to f }.collect { (r, f) ->
                _state.update { it.copy(localRecent = r.take(12), favoritesRow = f.take(12)) }
            }
        }
    }

    fun refresh() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            val mainResult = runCatching { repository.firstFeedWithBlocks() }
            mainResult.fold(
                onSuccess = { feed ->
                    val (hero, spotlight, freshRow) = digestMain(feed)
                    _state.update {
                        it.copy(
                            isLoading = false,
                            error = null,
                            hero = hero,
                            spotlight = spotlight,
                            feedRecent = feed.recentGames.take(12),
                            // Initial rows = "Fresh today" only; per-category rows
                            // arrive incrementally as fan-out finishes.
                            genreRows = listOfNotNull(freshRow),
                        )
                    }
                    fanOutGenreRows(excludeAppId = hero?.appId, prefix = listOfNotNull(freshRow))
                },
                onFailure = { err ->
                    _state.update { it.copy(isLoading = false, error = err.message) }
                },
            )
            refreshProfile()
        }
    }

    fun toggleFavorite(game: Game) {
        viewModelScope.launch { runCatching { repository.toggleFavorite(game) } }
    }

    private fun refreshProfile() {
        viewModelScope.launch {
            val p = runCatching { repository.userProfileWithRetry() }.getOrNull()
            if (p != null) _state.update { it.copy(profile = p) }
        }
    }

    /// Picks Hero + Spotlight + a "Fresh today" row from the main feed.
    /// Genre rows come from `fanOutGenreRows` because mobile-platform feeds
    /// don't carry titled `categorized` blocks.
    private fun digestMain(feed: FeedWithBlocks): Triple<Game?, SpotlightBlock?, GenreRow?> {
        fun isPromo(b: FeedBlock) =
            b.type.equals("promo", ignoreCase = true) || b.type.equals("adv", ignoreCase = true)
        val heroBlock = feed.blocks.firstOrNull { !isPromo(it) && it.items.isNotEmpty() }
        val hero = heroBlock?.items?.firstOrNull() ?: feed.flatGames.maxByOrNull { it.ratingCount }
        val spotlightBlock = feed.blocks.firstOrNull {
            !isPromo(it) && it !== heroBlock && it.items.size >= 5
        }
        val spotlight = spotlightBlock?.let {
            SpotlightBlock(it.title.ifEmpty { "Featured" }, it.items)
        }
        val freshRow = heroBlock?.items?.drop(1)?.takeIf { it.isNotEmpty() }?.let {
            GenreRow(title = "Fresh today", categoryName = null, games = it)
        }
        return Triple(hero, spotlight, freshRow)
    }

    /// Fans out one feed call per top category and merges results into
    /// `genreRows`. Done in parallel; results are reordered to match the
    /// original category list before being published.
    private suspend fun fanOutGenreRows(excludeAppId: Long?, prefix: List<GenreRow>) {
        val categories = runCatching { repository.categories() }.getOrNull().orEmpty()
        if (categories.isEmpty()) return
        val pick = categories.take(categoryRowLimit)
        val rows: List<Pair<GameCategory, List<Game>>> = coroutineScope {
            pick.map { cat ->
                async {
                    val feed = runCatching { repository.firstFeedWithBlocks(tab = cat.name) }.getOrNull()
                    val items = (feed?.flatGames.orEmpty()).filter { it.appId != excludeAppId }
                    cat to items.take(15)
                }
            }.awaitAll()
        }
        val ordered = rows
            .filter { it.second.isNotEmpty() }
            .map { (cat, items) -> GenreRow(title = cat.title, categoryName = cat.name, games = items) }
        _state.update { it.copy(genreRows = prefix + ordered) }
    }
}
